### DISCUSSION TITLE
Error while compiling modified version of DAFoam

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#373] [Installation]
Hi,

I am trying to make some modifications to the existing DAFoam code to contribute to it.

I want to use the DASpalartAllmarasFv3FieldInversion.C file and make some changes in it.

First, I went about and made those changes in the same file (I know I should create a separate file and make these changes but as a first step I wanted to check if I can recompile after making the changes).

I go to /home/<user>/dafoam/repos/dafoam-3.0.3 and do ./Allmake and I get this error

![modified_src_dafoam_erro](https://user-images.githubusercontent.com/10186475/198683721-e1884872-9e95-4909-bada-3f8f8e633888.png)

Now I know it is to do with codipack. But what seems to be the mistake that I am making here?

I use Ubuntu 20.04 LTS with dafoam v3.0.3. Everything compiled fine for me except the aerostruct case. I ran the periodic hill test case and the NACA0012 airfoil cases and they gave good results.

Any help will be appreciated.

**Comment by uttamcadambi07** [comment]
I am going to post the detailed error here. 

Before that, here are the steps that I did.

1. go into /home/<user>/dafoam/repos/dafoam-3.0.3/src/adjoint/DAModel/DATurbulenceModel/DASpalartAllmarasFv3FieldInversion.C
2. Make my modifications (I made some changes in void DASpalartAllmarasFv3FieldInversion::calcResiduals(const dictionary& options) function)
3. Go to /home/uttamcadambi/dafoam/repos/dafoam-3.0.3 and do ./Allclean and then ./Allmake
4. source $HOME/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/etc/bashrc && \
./Allclean && ./Allmake && \
source $HOME/dafoam/OpenFOAM/OpenFOAM-v1812-ADF/etc/bashrc && \
./Allclean && ./Allmake && \
pip install .

I tried running the periodic hill case and got this error 
![modified_src_dafoam_case_error](https://user-images.githubusercontent.com/10186475/198708686-190a5e21-b66b-4f3e-8b67-cfe9e8ed1f30.png)

Did I miss something?

**Reply by friedenhe** [reply]
It is likely that the codes you added had problems and the code did not compile properly (I think you should see an error message). You need to roll back your change and make sure you can compile and run the original version. Then, add your change step-by-step to figure out which change breaks the code.

**Reply by uttamcadambi07** [reply]
Thank you Dr. He. 
I will try what you suggested.
But is the procedure correct? I mean, is the order of making and compilation that I have followed correct?

**Reply by friedenhe** [reply]
First, go to the DAFoam folder, and then run the following commands to compile. 
<pre>
. $HOME/dafoam/loadDAFoam.sh && \
export DAFOAM_NO_WARNINGS=1 && \
./Allmake && \
source $HOME/dafoam/OpenFOAM/OpenFOAM-v1812-ADR/etc/bashrc && \
./Allclean && ./Allmake
</pre>

You don't need to compile the forward mode. **The key is to source loadDAFoam.sh before compiling**. Note that loadDAFoam.sh is not just sourcing the OpenFOAM environment, like foamenv, it also set many other environmental variables.

**Reply by uttamcadambi07** [reply]
Ah I understood now.
Compilation of forward mode is not needed because we use only the ADR mode to solve our adjoint equations in the Jacobian-free manner. The reverse mode AD in this case avoids having to compute and store Jacobians which the forward mode AD does. I hope my understanding is correct.

Your advice worked. I have managed to compile and my test case is running. I will keep this thread updated on the results. 

By the way, in my logOpt.txt file, I am getting this repeatedly:

Time = 9700

U Initial residual: (5.440786508439362e-11 4.284719554939937e-11 0.03538095951080856)
U   Final residual: (4.900575757327368e-12 1.592832920616493e-12 0.002797202011318399)
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371985747244
p Initial residual: 1.259715336676026e-10
p   Final residual: 1.25889796686841e-11
Time step continuity errors : sum local = 1.018467604668296e-15
                                 global = -1.132381497569258e-17
                             cumulative = -1.036763629061523e-15
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371985768564
nuTilda Initial residual: 1.25075036796183e-14
          Final residual: 8.099502153930112e-16
Bounding nuTilda>1e-16
yPlus min: 1.563956930174829 max: 2.497964821782095 mean: 2.133077886078058
FI-Ux-fieldInversion: 9.527043389912627
FI-beta-fieldInversion: 0
ExecutionTime = 478.06 s  ClockTime = 609 s

Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371985768564
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371985768564
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371985736584
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371985821864
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371985768564
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371985768564
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371985779224
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371985832524
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371985789885
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371985885825
Pressure gradient source: uncorrected Ubar = 0.02018800000000008, pressure gradient = 0.0007241371985629984
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371985864505
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371985747245
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371985875165
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371985768565
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371985853845
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371985832525
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371985853845
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371985896484
Pressure gradient source: uncorrected Ubar = 0.02018799999999997, pressure gradient = 0.0007241371985949785
Pressure gradient source: uncorrected Ubar = 0.02018800000000006, pressure gradient = 0.0007241371985757905
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371985832525
Pressure gradient source: uncorrected Ubar = 0.02018799999999995, pressure gradient = 0.0007241371985981766
Pressure gradient source: uncorrected Ubar = 0.02018799999999997, pressure gradient = 0.0007241371985928465
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371985843185
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371985949785
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371985896484
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371985917805
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371985907145
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371985971106
Pressure gradient source: uncorrected Ubar = 0.02018800000000005, pressure gradient = 0.0007241371985832526
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371985992425
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371985896485
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371985939125
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986003085
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371985971105
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986035065
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371985981764
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986013744
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371985928464
Pressure gradient source: uncorrected Ubar = 0.02018799999999993, pressure gradient = 0.0007241371986141664
Pressure gradient source: uncorrected Ubar = 0.02018799999999995, pressure gradient = 0.0007241371986077704
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371985960445
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986045725
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371985992424
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986067045
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986035065
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986067045
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986013744
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986067045
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986013744
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986099024
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986056385
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986099024
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986077704
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986152325
Pressure gradient source: uncorrected Ubar = 0.02018800000000006, pressure gradient = 0.0007241371985960445
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986067045
Pressure gradient source: uncorrected Ubar = 0.02018799999999996, pressure gradient = 0.0007241371986184304
Pressure gradient source: uncorrected Ubar = 0.02018799999999997, pressure gradient = 0.0007241371986173644
Pressure gradient source: uncorrected Ubar = 0.02018800000000005, pressure gradient = 0.0007241371986024404
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986152325
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986162985
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986152325
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986088364
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986162985
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986162985
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986194964
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986141664
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986162985
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986216285
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986173644
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986205624
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986226945
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986205625
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986280245
Pressure gradient source: uncorrected Ubar = 0.02018800000000006, pressure gradient = 0.0007241371986109685
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986280245
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371986162986
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986216285
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986269585
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986269585
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986258926
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986280245
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986258926
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986269585
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986269585
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986301565
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986226945
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986333545
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986237605
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986312225
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986365525
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986365525
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986301565
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986269585
Pressure gradient source: uncorrected Ubar = 0.02018799999999996, pressure gradient = 0.0007241371986397505
Pressure gradient source: uncorrected Ubar = 0.02018799999999996, pressure gradient = 0.0007241371986408165
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371986301565
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986397505
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371986290905
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986333545
Pressure gradient source: uncorrected Ubar = 0.02018799999999997, pressure gradient = 0.0007241371986418825
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986386845
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986365525
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986397505
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986450806
Pressure gradient source: uncorrected Ubar = 0.02018799999999997, pressure gradient = 0.0007241371986482785
Pressure gradient source: uncorrected Ubar = 0.02018800000000006, pressure gradient = 0.0007241371986301565
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371986376185
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986440146
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986450806
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986397505
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986461465
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986472125
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986450806
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986450806
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986461465
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986482785
Pressure gradient source: uncorrected Ubar = 0.02018799999999997, pressure gradient = 0.0007241371986546745
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371986429486
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986472125
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986504105
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986504105
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986504105
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986514765
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986450804
Pressure gradient source: uncorrected Ubar = 0.02018799999999997, pressure gradient = 0.0007241371986600045
Pressure gradient source: uncorrected Ubar = 0.02018800000000006, pressure gradient = 0.0007241371986418825
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986525425
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986536085
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986557404
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986514765
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986589384
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986514764
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986557404
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986568064
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986621365
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986546744
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986600045
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986610705
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986557405
Pressure gradient source: uncorrected Ubar = 0.02018799999999995, pressure gradient = 0.0007241371986706646
Pressure gradient source: uncorrected Ubar = 0.02018799999999997, pressure gradient = 0.0007241371986642685
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986589385
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986642685
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986621366
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986653345
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986600045
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986664005
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986642685
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986695985
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986653345
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986674665
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986674665
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986685325
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986674665
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986717305
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986621365
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986706645
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986717305
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986770605
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986685325
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986695985
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986749285
Pressure gradient source: uncorrected Ubar = 0.02018799999999997, pressure gradient = 0.0007241371986781265
Pressure gradient source: uncorrected Ubar = 0.02018800000000005, pressure gradient = 0.0007241371986642685
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986770605
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986759945
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986834565
Pressure gradient source: uncorrected Ubar = 0.02018800000000005, pressure gradient = 0.0007241371986695986
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986749285
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986759945
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986802586
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986759946
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986855886
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371986727967
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986802586
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986759946
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986813246
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986813246
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986845225
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986855885
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986834565
Pressure gradient source: uncorrected Ubar = 0.020188, pressure gradient = 0.0007241371986834565
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986898526
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986866546
Pressure gradient source: uncorrected Ubar = 0.02018800000000002, pressure gradient = 0.0007241371986845225
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986909186
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986877205
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986855885
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986941165
Pressure gradient source: uncorrected Ubar = 0.02018800000000003, pressure gradient = 0.0007241371986855885
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986898526
Pressure gradient source: uncorrected Ubar = 0.02018799999999999, pressure gradient = 0.0007241371986919845
Pressure gradient source: uncorrected Ubar = 0.02018799999999998, pressure gradient = 0.0007241371986951826
Pressure gradient source: uncorrected Ubar = 0.02018800000000004, pressure gradient = 0.0007241371986834566
Pressure gradient source: uncorrected Ubar = 0.02018800000000001, pressure gradient = 0.0007241371986919847




where is the print statement for this line? I tried looking for it but could not find it. I feel it is printing out the statement in some loop but I am not able to identify which one.

**Reply by friedenhe** [reply]
It is an issue caused by fvOptions in the DAFoam tutorial. We have just updated the case and checkout our change from here: https://github.com/DAFoam/tutorials/commit/977616354f249f07bf810cbed353632460717a81. Bottom line: we don't support fvOptions for now.

**Reply by uttamcadambi07** [reply]
Ah okay thanks Dr. He.

We are adding a source term in the runscript.py in the form of a uniform pressure gradient. I got this from the periodic hill tutorial on: https://dafoam.github.io/mydoc_tutorials_field_inversion_ph.html. There it says:

![image](https://user-images.githubusercontent.com/10186475/199043221-be2711b8-5fa8-468a-829b-0be90d04599d.png)

I have two questions:
1. Do these lines given in the screenshot above that is included in the runscript.py make the fvOptions file redundant? (which is why you suggested the change of commenting out the contents of fvOptions)? Because the way I see it, the meanVelocityForce type in fvOptions helps us maintain a constant volume averaged velocity of Ubar = 0.020188. 
2. What does the comment in the screenshot above "#get this value by running the original OpenFOAM solver" mean? Does it mean: For the case of the periodic hill, the value given in the tutorials files can be used. For other cases where bulk velocity should be maintained constant such as say channel flow, we need to compute the value from a pre-cursor simulation and for cases where it is not needed, these lines can be removed. Is this understanding correct?

**Reply by friedenhe** [reply]
@obidar Can you answer this question?

**Comment by omid-bidar** [comment]
That's correct. For the tutorial case the fvSource value is generated by running a pre-cursor simulation using OpenFOAM flow solver, and the value is the `pressure gradient` in the log file (after the flow has converged sufficiently). For a different geometry and/or Reynolds number we will need to repeat this process and get the correct  pressure gradient value. In cases where you do not need to set a constant volume averaged velocity, you can remove the fvSource dictionary from the runScript.py, as mentioned in the tutorial.

**Reply by uttamcadambi07** [reply]
Hi Omid. I just wanted to revisit this thread to clarify something.
"For the tutorial case the fvSource value is generated by running a pre-cursor simulation using OpenFOAM flow solver, and the value is the pressure gradient in the log file (after the flow has converged sufficiently)."

So for a periodic hill, since it is a channel, you would have \tau_w = -dp/dx. So after I finish my simulation, I estimate the value of \tau_w and this will give me my dp/dx. Is that what you meant by this statement? 

Secondly, the channel height is 3.035H where H = 1. If I need a bulk Reynolds number of 5600, I would have to set the meanVelocityForce (or in this case, gradP) as a source term to ensure that my bulk velocity is 0.020188. So I can calculate the Reynolds number as Ub*H/(\nu_\infty). Plugging in the values = 0.02*1/5e-06 (these are the values that are there in the kEpsilon.tar.gz file in the dafoam tutorials) = I get 4000 as my Reynolds number. So my guess is that the gradP that you set does not correspond to the same as maintaining a constant Ub of 0.020188. Please let me know if my understanding on this is correct.

**Reply by uttamcadambi07** [reply]
Right, I just made the computation.

By setting the Ubar = (0.020188 0 0) in fvoptions during a simulation of the kEpsilon case, I computed the \tau_w to be 6.61884215964998e-06 and the gradP that you set in runScript_v2.py as daOptions is 6.634074021107811e-06. I think I have answered my own question but for others - since fvOptions is not supported by DAFoam as mentioned before, for a channel flow case, compute the wallShearStress on upper and lower walls, add them up and add it as gradP in the daOptions of the runScript. 

So in the order of steps:
1. Determine bulk Reynolds number. 
2. Estimate the Ub that you wish to maintain.
3. set that in fvOptions using meanVelocityForce for the pre-cursor simulation
4. estimate nu_\infty based on Re_b, H (=1) and Ub (which you set)
5. Run the simulation
6. Compute top and bottom wall shear stresses and add them up
7. This is the gradP that you set in the runScript when doing the field inversion.
